src/query/parser/promql/parse.go

```diff
@@ -228,11 +229,37 @@ func (p *parseState) walk(node pql.Node) error {
 			return nil
 		}
 
-		expressions := n.Args
-		argTypes := n.Func.ArgTypes
-		argValues := make([]interface{}, 0, len(expressions))
-		stringValues := make([]string, 0, len(expressions))
-		for i, argType := range argTypes {
+		var (
+			// argTypes describes Prom's expected argument types for this call.
+			argTypes = n.Func.ArgTypes
+			// expressions describes the actual arguments for this call.
+			expressions = n.Args
+			argCount    = len(argTypes)
+			exprCount   = len(expressions)
+			numVals     = argCount
+			variadic    = n.Func.Variadic
+		)
+
+		if variadic == 0 {
+			if argCount != exprCount {
+				return fmt.Errorf("incorrect number of expressions(%d) for %q, "+
+					"received %d", exprCount, n.Func.Name, argCount)
+			}
+		} else {
+			if argCount-1 > exprCount {
+				return fmt.Errorf("incorrect number of expressions(%d) for variadic "+
+					"function %q, received %d", exprCount, n.Func.Name, argCount)
+			}
+
+			if argCount != exprCount {
+				numVals--
```

Comment:
Maybe we should call this `numExpVals` or something to reflect it's what we're expecting?